x-pack/plugins/fleet/server/services/secrets.test.ts

```diff
@@ -752,50 +794,125 @@ describe('diffSecretPaths', () => {
+        toCreate: [
+          {
+            path: 'somepath2',
+            value: { value: 'newvalue' },
+          },
+        ],
+        toDelete: [],
+        noChange: [paths1[0]],
+      });
     });
   });
 
-  it('single secret added', () => {
-    const paths1 = [
-      {
-        path: 'somepath1',
-        value: {
-          value: {
-            isSecretRef: true,
-            id: 'secret-1',
-          },
+  describe('extractAndWriteSecrets', () => {
+    const esClientMock = elasticsearchServiceMock.createInternalClient();
+
+    esClientMock.transport.request.mockImplementation(async (req) => {
+      return {
+        id: uuidv4(),
+      };
+    });
+
+    beforeEach(() => {
+      esClientMock.transport.request.mockClear();
+    });
+
+    const mockIntegrationPackage = {
+      name: 'mock-package',
+      title: 'Mock package',
+      version: '0.0.0',
+      description: 'description',
+      type: 'integration',
+      status: 'not_installed',
+      vars: [
+        { name: 'pkg-secret-1', type: 'text', secret: true, required: true },
+        { name: 'pkg-secret-2', type: 'text', secret: true, required: false },
+      ],
+      data_streams: [
+        {
+          dataset: 'somedataset',
+          streams: [
+            {
+              input: 'foo',
+              title: 'Foo',
+            },
+          ],
         },
-      },
-    ];
-
-    const paths2 = [
-      paths1[0],
-      {
-        path: 'somepath2',
-        value: { value: 'newvalue' },
-      },
-    ];
-
-    expect(diffSecretPaths(paths1, paths2)).toEqual({
-      toCreate: [
+      ],
+      policy_templates: [
         {
-          path: 'somepath2',
-          value: { value: 'newvalue' },
+          name: 'pkgPolicy1',
+          title: 'Package policy 1',
+          description: 'test package policy',
+          inputs: [
+            {
+              type: 'foo',
+              title: 'Foo',
+              vars: [],
+            },
+          ],
         },
       ],
-      toDelete: [],
-      noChange: [paths1[0]],
+    } as unknown as PackageInfo;
+
+    describe('when only required secret value is provided', () => {
+      it('returns single secret reference for required secret', async () => {
+        const mockPackagePolicy = {
+          vars: {
+            'pkg-secret-1': {
+              value: 'pkg-secret-1-val',
+            },
+          },
+          inputs: [],
+        } as unknown as NewPackagePolicy;
+
+        const result = await extractAndWriteSecrets({
+          packagePolicy: mockPackagePolicy,
+          packageInfo: mockIntegrationPackage,
+          esClient: esClientMock,
+        });
+
+        expect(esClientMock.transport.request).toHaveBeenCalledTimes(1);
+        expect(result.secretReferences).toHaveLength(1);
+      });
+    });
+
+    describe('when both required and optional secret values are provided', () => {
+      it('returns secret reference for both required and optional secret', async () => {
```

Comment:
nit: you could add the when to the testcase itself e.g. `should return secret reference ... when both ....`